openmls/src/group/public_group/staged_commit.rs

```diff
@@ -0,0 +1,260 @@
+    ///  - Verifies the confirmation tag/membership tag
+    /// Returns a [`StagedCommit`] that can be inspected and later merged into
+    /// the group state either with [`CoreGroup::merge_commit()`] or
+    /// [`PublicGroup::merge_diff()`] This function does the following checks:
+    ///  - ValSem101
+    ///  - ValSem102
+    ///  - ValSem104
+    ///  - ValSem105
+    ///  - ValSem106
+    ///  - ValSem107
+    ///  - ValSem108
+    ///  - ValSem110
+    ///  - ValSem111
+    ///  - ValSem112
+    ///  - ValSem200
+    ///  - ValSem201
+    ///  - ValSem202: Path must be the right length
+    ///  - ValSem203: Path secrets must decrypt correctly
+    ///  - ValSem204: Public keys from Path must be verified and match the
+    ///               private keys from the direct path
+    ///  - ValSem205
+    ///  - ValSem240
+    ///  - ValSem241
+    ///  - ValSem242
+    ///  - ValSem243
+    ///  - ValSem244
+    /// Returns an error if the given commit was sent by the owner of this
+    /// group.
+    /// TODO #1255: This will be used by the `process_message` function of the
+    /// `PublicGroup` later on.
+    #[allow(unused)]
+    pub(crate) fn stage_commit(
+        &self,
+        mls_content: &AuthenticatedContent,
+        proposal_store: &ProposalStore,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<StagedCommit, StageCommitError> {
+        let ciphersuite = self.ciphersuite();
+
+        let (commit, proposal_queue, sender_index) =
+            self.validate_commit(mls_content, proposal_store, backend)?;
+
+        // Create the provisional public group state (including the tree and
+        // group context) and apply proposals.
+        let mut diff = self.empty_diff();
+
+        let apply_proposals_values = diff.apply_proposals(&proposal_queue, None)?;
+
+        // Determine if Commit has a path
+        if let Some(path) = commit.path.clone() {
+            // Update the public group
+            // ValSem202: Path must be the right length
+            diff.apply_received_update_path(backend, ciphersuite, sender_index, &path)?;
+        } else if apply_proposals_values.path_required {
+            // ValSem201
+            return Err(StageCommitError::RequiredPathNotFound);
+        };
+
+        // Update group context
+        diff.update_group_context(backend)?;
+
+        // Update the confirmed transcript hash before we compute the confirmation tag.
+        diff.update_confirmed_transcript_hash(backend, mls_content)?;
+
+        let received_confirmation_tag = mls_content
+            .confirmation_tag()
+            .ok_or(StageCommitError::ConfirmationTagMissing)?;
+
+        // If we have private key material, derive the secrets for the next
+        // epoch and check the confirmation tag.
+        diff.update_interim_transcript_hash(
+            ciphersuite,
+            backend,
+            received_confirmation_tag.clone(),
+        )?;
```

Comment:
All of this could be moved to a separate function that's creating the diff to scope the `mut` and separate the operations on the diff from the commit staging.
Same for the `stage_commit` in the core group. This way it's also easier to see where the two differ and making sure we keep them in sync.
Can you also add comments on both functions that the other one exists?